### PR TITLE
vec: support RVV

### PR DESCRIPTION
**Pull request by zhangfeiv0**
# Feature
SIMD vectorization for RISCV64 architecture

implemented types:
- float
- bf16
- qint8
- quint8
- qint32

The implementation of the above types can support compiling PyTorch on the RISC-V architecture. For other types, we hope these codes can be reviewed, and then we proceed to implement other types.

# Vector Length Issues
Currently, PyTorch covers different SIMD extensions on various platforms, such as AVX2, AVX512, NEON, VSX, and ZARCH. All the above arch is designed with fixed register size. Nowadays, there are also some vector ISA extensions is designed with size-less instruction, such as RISC-V Vector (aka. RVV) and ARM SVE/SVE2. Taking RVV as an example, different devices that support the RVV extension may have different vector register sizes, any size powered of 2 && between 128 bits and 65536 bits is legally for a RVV application processor.

The RISC-V vector extension supports variable-length vectors using an explicit vector length (`vl`) value set by applications in a configuration register, up to a hardware-specific maximum length of VLEN bits. In the C language vector intrinsics on which our port is based, each intrinsic takes an explicit vector length parameter. 

For the sake of generality, we have simply hardwired this vector length to a value based on `__riscv_v_min_vlen` (128 bits). When we try to wrapper a rvv native type in a struct or a class, clang will report error: `field has sizeless type 'vfloat32m1_t' (aka '__rvv_float32m1_t')`.

# Build
Clang version: 17.0.2
Platform: CanMV-K230、QEMU
Architecture: riscv64
PyTorch version: main

`export USE_CUDA=0 USE_DISTRIBUTED=0 USE_MKLDNN=0 MAX_JOBS=4 CMAKE_CXX_COMPILER=clang++ CMAKE_C_COMPILER=clang`
`python3 setup.py develop --cmake`

# Correctness Test
1. test on QEMU for vector register widths of 128 bits, 256 bits, 512 bits, and 1024 bits.
<details> <summary> QEMU Test Result</summary>

`qemu-system-riscv64 -cpu rv64,v=on,vlen=128 ......`
`./vec_test_all_types_RVV`

`qemu-system-riscv64 -cpu rv64,v=on,vlen=256 ......`
`./vec_test_all_types_RVV`

`qemu-system-riscv64 -cpu rv64,v=on,vlen=512 ......`
`./vec_test_all_types_RVV`

`qemu-system-riscv64 -cpu rv64,v=on,vlen=1024 ......`
`./vec_test_all_types_RVV`

</details>

2. test on the CanMV-K230 hardware, which supports the RISC-V Vector 1.0 standard.
<details> <summary>CanMV-K230 Test Result</summary>

`./pytorch/build/bin/vec_test_all_types_RVV`
RVV

</details>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by zhangfeiv0** [comment]
@malfet @ezyang

**Comment by ezyang** [comment]
We likely have a reviewer problem; it's not clear who on the core team should be reviewing RISCV64

**Comment by zhangfeiv0** [comment]
> We likely have a reviewer problem; it's not clear who on the core team should be reviewing RISCV64

Hello，@ezyang
Do you have any plans to discuss this issue and designate a maintainer or add additional maintainers to be responsible for the RISC-V architecture?

**Comment by ezyang** [comment]
It looks like this is for @malfet

**Comment by abhishek-iitmadras** [comment]
Hi @zhangfeiv0, Curious to know if there have been any perf improvements across ops or kernels. Do you have any data you can share? Thank you!

**Comment by abhishek-iitmadras** [comment]
@ezyang @malfet 
This PR involves non-trivial changes and should have a thorough CI plan. 
Do we have CI plan in place for such a PR?

**Comment by zhangfeiv0** [comment]
> Hi @zhangfeiv0, Curious to know if there have been any perf improvements across ops or kernels. Do you have any data you can share? Thank you!

Hello, we have noticed that support for SVE has been merged into PyTorch (#119571), and we are re-evaluating the implementation of RVV code, which requires some modifications. We will subsequently submit the code and performance data.

@malfet @ezyang @abhishek-iitmadras

**Comment by malfet** [comment]
>  This PR involves non-trivial changes and should have a thorough CI plan. Do we have CI plan in place for such a PR?

I don't think we have any CI (nor plans for it) to start supporting RISC V. But if it's readily available in cloud, please comment here and let's talk about potential for enabling something like that

**Comment by wangfanv0** [comment]
> > This PR involves non-trivial changes and should have a thorough CI plan. Do we have CI plan in place for such a PR?
> 
> I don't think we have any CI (nor plans for it) to start supporting RISC V. But if it's readily available in cloud, please comment here and let's talk about potential for enabling something like that

In our PyTorch RVV development, we primarily rely on QEMU for compilation. We use QEMU 8 (or QEMU 9) to launch an operating system on the host machine (Ubuntu or Windows), and then perform the PyTorch compilation within this operating system. The operating system can be Ubuntu (we are using openEuler), with the latest kernel built to support RVV1.0. Is it possible to build a Docker image like this for CI testing?

**Comment by zhangfeiv0** [comment]
To my knowledge, there is currently no cloud service that supports RVV. In some software packages, CI is achieved by cross-compiling for the RISC-V architecture and using QEMU for testing.

in sleef ci:

```yaml
- name: Setup QEMU
        uses: docker/setup-qemu-action@v3.0.0
        with:
          platforms: ${{ matrix.binfmt || matrix.arch }}

......

- name: Test ${{ matrix.arch }}
        env:
          CTEST_OUTPUT_ON_FAILURE: "TRUE"
        run: |
          if [[ -n "${{ matrix.qemu_cpu }}" ]]; then
            export QEMU_CPU="${{ matrix.qemu_cpu }}"
          fi
          cd _build-${{ matrix.arch }}
          ctest -j$(nproc)
```

in opencv ci log:

```shell
Run cd $HOME/build
  cd $HOME/build
  /opt/riscv/bin/qemu-riscv64 -cpu rv64,v=true,vext_spec=v1.0 -L /opt/riscv/sysroot bin/opencv_test_core \
    --skip_unstable=1 \
    --gtest_filter=*
```

We can use a method similar to [docker/setup-qemu-action](https://github.com/docker/setup-qemu-action) to run RISC-V Docker images on an x86_64 host and build PyTorch. Here is an example of testing this PR on an x86_64 host:

```shell
# In x86_64 host
# Setup QEMU, similar to docker/setup-qemu-action
sudo docker pull tonistiigi/binfmt:latest
sudo docker run --rm --privileged tonistiigi/binfmt:latest --install riscv64

# Import ubuntu24.04-riscv64
wget https://cdimage.ubuntu.com/ubuntu-base/releases/24.04/release/ubuntu-base-24.04.1-base-riscv64.tar.gz
sudo docker import ubuntu-base-24.04.1-base-riscv64.tar.gz ubuntu:24.04
sudo docker run -it ubuntu:24.04 /bin/bash

# In ubuntu:24.04 container
# Install common dependencies, Some of these may not be necessary
apt update
export DEBIAN_FRONTEND=noninteractive
apt install -y llvm-17 clang-17 git libopenblas-dev libblas-dev m4 cmake cython3 ccache python3-pip
update-alternatives --install /usr/bin/clang clang /usr/bin/clang-17 50
update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-17 50
pip install typing_extensions  --break-system-packages
pip install markupsafe --break-system-packages

# test for this pr in ubuntu:24.04 container
cd /root
git clone https://github.com/zhangfeiv0/pytorch.git -b vec-rvv
cd pytorch
git submodule sync
git submodule update --init --recursive

# enable RVV, vlen=128
export QEMU_CPU="rv64,v=true,vlen=128,vext_spec=v1.0"
# flags
export USE_CUDA=0 USE_DISTRIBUTED=0 USE_MKLDNN=0 MAX_JOBS=$(nproc --ignore=4) CMAKE_CXX_COMPILER=clang++ CMAKE_C_COMPILER=clang
# compile
python3 setup.py develop

# test for vec lib
./build/bin/vec_test_all_types_RVV
```

Perhaps, the above method could be used for RISC-V-related PR validation, or could it help us integrate the above process into pytorch CI? @malfet 


Additionally, attach the performance data on the K230 hardware.
[test_operators.txt](https://github.com/user-attachments/files/17470486/test_operators.txt)


|                | -O3<br />(scalar) | -O3 -march=rv64gcv<br />（auto vectorization） | -O3 -march=rv64gcv<br />（this pr，use RVV intrinsics） |
| -------------- | ----------------: | ---------------------------------------------: | ------------------------------------------------------: |
| torch.sin            |          11.411 s |                                       11.173 s |                                                 4.825 s |
| torch.cos            |          10.179 s |                                          9.9 s |                                                 5.226 s |
| torch.tan            |          19.929 s |                                       20.737 s |                                                 6.928 s |
| torch.round          |           8.726 s |                                        7.435 s |                                                 2.024 s |
| torch.sqrt           |           3.211 s |                                        1.188 s |                                                 1.137 s |
| torch.pow            |           3.115 s |                                        0.368 s |                                                  0.36 s |
| torch.sum            |           1.953 s |                                        0.305 s |                                                 0.268 s |
| torch.frac           |           4.801 s |                                        5.721 s |                                                 0.734 s |
| torch.mul            |           4.423 s |                                        0.408 s |                                                 0.411 s |
| torch.div            |           4.549 s |                                        1.273 s |                                                 1.212 s |

With this PR, we introduced RVV and we did observe performance improvements in ops like add, sub, mul, div, pow, sqrt, round, tan, cos, sin, etc.

**Comment by zhangfeiv0** [comment]
@malfet

**Comment by zhangfeiv0** [comment]
Pinging again ... Any help I can get on this?
@malfet

**Comment by camel-cdr** [comment]
Correct me if I'm wrong, but it currently seems to build only for a specific vector length, which is detected at build time.
Since you set `-mrvv-vector-bits`, the compiler will generate and auto vectorize code in a vector length specific way, which may break even when running code on a larger VLEN than it was compiled with.
I'm not sure about the distribution model of pytorch, but this makes it almost impossible to properly distribute binaries.
Would it be possible to build the RVV backend for multiple VLEN (let's say 128 to 1024) and runtime select the correct one?

**Comment by zhangfeiv0** [comment]
> Correct me if I'm wrong, but it currently seems to build only for a specific vector length, which is detected at build time. Since you set `-mrvv-vector-bits`, the compiler will generate and auto vectorize code in a vector length specific way, which may break even when running code on a larger VLEN than it was compiled with. I'm not sure about the distribution model of pytorch, but this makes it almost impossible to properly distribute binaries. Would it be possible to build the RVV backend for multiple VLEN (let's say 128 to 1024) and runtime select the correct one?

We have adopted a method similar to the SVE ISA ([sve_helper.h](https://github.com/pytorch/pytorch/blob/fe4fa1df9fa981f31f85170e40a754479759267f/aten/src/ATen/cpu/vec/sve/sve_helper.h#L10)), detecting the vector bit width during build time, but in actual code, it only controls 256-bit operations, and the operations in the vec lib are also implemented for 256 bits.

Setting `-mrvv-vector-bits` indeed causes the program to fail on machines with a larger VLEN than at compile time, so for larger VLEN machines, recompilation is required, which is a significant drawback. Regarding this issue, perhaps using arrays in Vectorized to store, combined with `__riscv_v_min_vlen` and `m2` to control only 256 bits, could solve the compatibility issue, but there would be a performance loss for larger VLEN.

When we try to wrap RVV native types in structures or classes, clang reports an error: `field has sizeless type 'vfloat32m1_t' (aka '__rvv_float32m1_t')`, which is why we use `-mrvv-vector-bits` to specify a particular vector length, which can better integrate with the current vec lib.

Currently, there is no good method for a single RVV backend to be shared from 128 to 1024 bit widths, and to correctly control the vector length at runtime through something like `__riscv_vsetvlmax_e32m2()`.If you have any good solutions, please feel free to propose them as well.

**Comment by camel-cdr** [comment]
I had an idea for a solution: https://github.com/riscv-non-isa/rvv-intrinsic-doc/issues/378

But I'm not sure it's the best way of doing things. It certainly requires compiler support.

**Comment by zhangfeiv0** [comment]
Thank you for your reply. As I understand it, the approach would be to define a sufficiently large buffer within a class or structure, such as `float values[32]`, to meet the VLEN requirement of 128 to 1024 lengths, and then use `__riscv_vsetvlmax_xxx()` to operate with an appropriate length at runtime. Besides the redundant load/store mentioned in [riscv-non-isa/rvv-intrinsic-doc#378](https://github.com/riscv-non-isa/rvv-intrinsic-doc/issues/378), this large buffer would cause space waste for machines with VLEN less than 1024, although it can achieve RVV backend compatibility for VLEN ranging from 128 to 1024.

Currently, PyTorch lacks CI testing for RISC-V, and there is a lack of sufficient discussions like the one mentioned above, which also prevents our PR from moving forward.
